File: backend/app/locales.py
"""Shared locale resolution for CalorieApp identity and UI surfaces."""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any, Iterator, Optional


REGISTRY_PATH = Path(__file__).with_name("data") / "locales.json"


@lru_cache(maxsize=1)
def locale_registry() -> dict[str, Any]:
    """Load the deployable copy of the frozen v1 locale registry."""
    try:
        value = json.loads(REGISTRY_PATH.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        raise RuntimeError("CalorieApp locale registry is unavailable") from exc
    if not isinstance(value, dict) or not isinstance(value.get("locales"), list):
        raise RuntimeError("CalorieApp locale registry is malformed")
    return value


def supported_locale_tags() -> tuple[str, ...]:
    return tuple(locale["tag"] for locale in locale_registry()["locales"])


@lru_cache(maxsize=1)
def _locale_identifier_map() -> dict[str, str]:
    identifiers: dict[str, str] = {}
    for locale in locale_registry()["locales"]:
        tag = locale["tag"]
        for identifier in [tag, *locale.get("aliases", [])]:
            identifiers[_normalize_identifier(identifier)] = tag
    return identifiers


def _normalize_identifier(value: str) -> str:
    return value.strip().replace("_", "-").lower()


def _requested_candidates(value: Optional[str]) -> Iterator[str]:
    if not value:
        return
    for part in value.split(","):
        candidate = part.split(";", 1)[0].strip()
        if candidate and candidate != "*":
            yield _normalize_identifier(candidate)


def resolve_locale(value: Optional[str]) -> str:
    """Resolve a locale tag or Accept-Language-like value with English fallback."""
    registry = locale_registry()
    identifiers = _locale_identifier_map()
    canonical_primary_tags = {
        locale["tag"].lower(): locale["tag"]
        for locale in registry["locales"]
        if "-" not in locale["tag"]
    }

    for candidate in _requested_candidates(value):
        exact = identifiers.get(candidate)
        if exact:
            return exact
        primary = candidate.split("-", 1)[0]
        if primary in canonical_primary_tags:
            return canonical_primary_tags[primary]
    return registry["fallback_locale"]


def locale_direction(value: Optional[str]) -> str:
    resolved = resolve_locale(value)
    for locale in locale_registry()["locales"]:
        if locale["tag"] == resolved:
            return locale["direction"]
    return "ltr"

File: frontend/lib/locales.ts
import localeRegistry from "@/config/locales.json";

export type LocaleDirection = "ltr" | "rtl";

export type SupportedLocaleDefinition = {
  tag: string;
  english_name: string;
  native_name: string;
  direction: LocaleDirection;
  source: boolean;
  aliases: string[];
};

export const sourceLocale = localeRegistry.source_locale;
export const fallbackLocale = localeRegistry.fallback_locale;
export const supportedLocales =
  localeRegistry.locales as SupportedLocaleDefinition[];

const identifierMap = new Map<string, string>();
const canonicalPrimaryTags = new Map<string, string>();

function normalizeIdentifier(value: string): string {
  return value.trim().replace(/_/g, "-").toLowerCase();
}

for (const locale of supportedLocales) {
  identifierMap.set(normalizeIdentifier(locale.tag), locale.tag);
  for (const alias of locale.aliases) {
    identifierMap.set(normalizeIdentifier(alias), locale.tag);
  }
  if (!locale.tag.includes("-")) {
    canonicalPrimaryTags.set(locale.tag.toLowerCase(), locale.tag);
  }
}

export function resolveLocale(value?: string | null): string {
  if (value) {
    for (const part of value.split(",")) {
      const rawCandidate = part.split(";", 1)[0]?.trim();
      if (!rawCandidate || rawCandidate === "*") {
        continue;
      }
      const candidate = normalizeIdentifier(rawCandidate);
      const exact = identifierMap.get(candidate);
      if (exact) {
        return exact;
      }
      const primary = candidate.split("-", 1)[0];
      const supportedPrimary = canonicalPrimaryTags.get(primary);
      if (supportedPrimary) {
        return supportedPrimary;
      }
    }
  }
  return fallbackLocale;
}

export function localeDirection(value?: string | null): LocaleDirection {
  const resolved = resolveLocale(value);
  return (
    supportedLocales.find((locale) => locale.tag === resolved)?.direction ??
    "ltr"
  );
}
